src/components/ApiHistoryTable.tsx

```diff
@@ -49,12 +55,20 @@ function generateTableRow(
       >
         {/* Semver shenanigans, feature backported to both ^7.1.0 and ^6.3.0 would not be present in 7.0.0 */}
         <pre>
-          {release === version ? '>=' : '^'}
+          {release === version || isBackportMajor ? '>=' : '^'}
           {version}
         </pre>
       </a>
     );
-  });
+
+    formattedVersions.push(formattedVersion);
+
+    // If backport is a major (i.e. >=x.0.0), no need to include the release or any other backports
+    if (isBackportMajor) break;
```

Comment:
Is skipping "any other backports" correct behavior here? I think we still want to display the case where it was backported to an older branch, while also consolidating the backport majors.

E.g., if it landed on `main`, then was backported to `31-x-y` while that branch was still in beta, and also backported to `30-x-y`, the release versions for the PR might look like 32.0.0 (main release), 31.0.0 (backport release), and 30.x.y. In this case we'd want the API history to display: ">=31.0.0, ^30.x.y". I think with breaking here it would only display ">=31.0.0"?

Reply:
Man, @MarshallOfSound, I like this feature and thank you for suggesting it. But I feel like I need a degree in theoretical physics just to implement/debug/discuss it correctly 😭.

@dsanders11 Due to the witchcraft/black magic necessary to implement this, I had to mess around with the sorting of the releases/backports. Are you sure you didn't miss these changes when you were trying to visualize the hell spawn that is this feature?

[`a047f6a` (#594)](https://github.com/electron/website/pull/594/commits/a047f6a2b6f09ee4d6a3950a95ff516643902a2d#diff-1dd7a747a324f6f8916a65d87d4c435b9041ed1b0a0d9fa48518423b4d3012edR39-R40)

(We now iterate from older to newer, i.e. 29, 30, etc. So, in your example, the `break;` would happen on `30.x.y`.

```patch
-  // Sort by major version number e.g. 30.0.0 -> 30 in descending order i.e. 30, 29, ...
-  allVersions.sort((a, b) => Number(b.split('.')[0]) - Number(a.split('.')[0]));
+  // Sort by major version number e.g. 30.0.0 -> 30 in ascending order i.e. 29, 30, ...
+  allVersions.sort((a, b) => Number(a.split('.')[0]) - Number(b.split('.')[0]));
```

[`a047f6a` (#594)](https://github.com/electron/website/pull/594/commits/a047f6a2b6f09ee4d6a3950a95ff516643902a2d#diff-1dd7a747a324f6f8916a65d87d4c435b9041ed1b0a0d9fa48518423b4d3012edR70-R71)

(We then reverse the order so the table display newest on the top i.e. 30, 29, ...)

```patch
+  // Reverse the order of the versions so that the latest version is first i.e. 30, 29, ...
+  formattedVersions.reverse();
```

Reply:
I think you're right @piotrpdev, it looks like it's working as intended, I think the comment is just a bit confusing. The way I originally read "no need to include the release" made it sound like it was not including the current release, but I now see it means "no need to include the **main** release". I think the comment would be easier to understand if it said something like "no need to check newer releases".

It looks like `semver` has a function that sounds like it would be helpful here, `simplifyRange`, but I tested it out and it doesn't produce the output we'd want. It seems to favor turning `^` into `x - y` style ranges.

Reply:
I played around with the code a bit and I think some restructuring can make it less confusing. 🙂

```diff
diff --git a/src/components/ApiHistoryTable.tsx b/src/components/ApiHistoryTable.tsx
index 728f133..9456df1 100644
--- a/src/components/ApiHistoryTable.tsx
+++ b/src/components/ApiHistoryTable.tsx
@@ -1,6 +1,7 @@
 import { Details } from '@docusaurus/theme-common/Details';
 import React from 'react';
 import ReactMarkdown from 'react-markdown';
+import * as semver from 'semver';
 
 import {
   ApiHistory,
@@ -28,47 +29,30 @@ function generateTableRow(
   const prNumber = Number(change['pr-url'].split('/').at(-1));
   const key = prNumber + '-' + type;
 
-  const allVersions: Array<string> = [];
+  // Map versions to semver syntax - use a caret unless it was released in
+  // an X.0.0 release, which gets >= since every release after that has it
+  const versionRanges: Array<string> = semver
+    .rsort(prReleaseVersions?.backports ?? [])
+    .map((version) => (version.endsWith('.0.0') ? '>=' : '^') + version);
 
+  // Only include the main release if it wasn't backported and released in
+  // an X.0.0 release. This consolidates ranges like >=30.0.0 || >=29.0.0
+  // into a single >=29.0.0 since that's more intuitive for developers.
   const release = prReleaseVersions?.release;
-  if (release) allVersions.push(release);
-
-  const backports = prReleaseVersions?.backports;
-  if (backports) allVersions.push(...backports);
-
-  // Sort by major version number e.g. 30.0.0 -> 30 in ascending order i.e. 29, 30, ...
-  allVersions.sort((a, b) => Number(a.split('.')[0]) - Number(b.split('.')[0]));
-
-  const formattedVersions: JSX.Element[] = [];
-
-  for (const version of allVersions) {
-    const [, minor, patch] = version.split('.');
-    const isBackportMajor =
-      release !== version && minor === '0' && patch === '0';
-
-    const formattedVersion = (
-      <a
-        key={version}
-        href={change['pr-url']}
-        target="_blank"
-        rel="noopener noreferrer"
-      >
-        {/* Semver shenanigans, feature backported to both ^7.1.0 and ^6.3.0 would not be present in 7.0.0 */}
-        <pre>
-          {release === version || isBackportMajor ? '>=' : '^'}
-          {version}
-        </pre>
-      </a>
-    );
-
-    formattedVersions.push(formattedVersion);
-
-    // If backport is a major (i.e. >=x.0.0), no need to include the release or any other backports
-    if (isBackportMajor) break;
+  if (release && !versionRanges.find((version) => version.startsWith('>='))) {
+    versionRanges.unshift(`>=${release}`);
   }
 
-  // Reverse the order of the versions so that the latest version is first i.e. 30, 29, ...
-  formattedVersions.reverse();
+  const formattedVersions = versionRanges.map((version) => (
+    <a
+      key={version}
+      href={change['pr-url']}
+      target="_blank"
+      rel="noopener noreferrer"
+    >
+      <pre>{version}</pre>
+    </a>
+  ));
 
   let changesJsx: JSX.Element | undefined;
 
```

Reply:
Patch applied in [`e61e7d6` (#594)](https://github.com/electron/website/pull/594/commits/e61e7d6fee589a7a4b6dcede8664bc867bee6875)